### PR TITLE
SELV3-546: Datasets definition

### PR DESCRIPTION
**Pull request by anawrotsoldevelo**
SELV3-548 - Create endpoints for adding new dataset, updating it and deleting
SELV3-546 - Implement possibility to store datasets
SELV3-547 - Add new table which stores datasets data

### CODE REVIEW

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+/**
+ * Controller used to expose Servers via HTTP.
+ */
+@Controller
+@RequestMapping(DatasetController.RESOURCE_PATH)
+@Transactional
+public class DatasetController extends BaseController {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java:60]: Is authorization (user needs a role to read/create/update/delete) missing or are they handled other way or it's left for later?

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public DatasetDto updateDataset(@PathVariable("id") UUID id, @RequestBody DatasetDto datasetDto) {
+    if (null != datasetDto.getId() && !Objects.equals(datasetDto.getId(), id)) {
+      throw new ValidationMessageException(MessageKeys.ERROR_DATASET_ID_MISMATCH);
+    }
+
+    LOGGER.debug("Updating dataset");
+    Dataset dataset;
+    Optional<Dataset> datasetOptional = datasetRepository.findById(id);
```
Comment (pwargulak) [inline src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java:141]: Or something like this, using Optional's methods:

```suggestion
    Dataset datasetToSave = datasetRepository.findById(id).map(dataset -> {
      dataset.updateFrom(datasetDto);
      return dataset;
    }).orElseGet(() -> Dataset.newInstance(datasetDto));
```

File: src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java [hunk 3]
```diff
@@ -57,21 +59,54 @@
 public class ServerController extends BaseController {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java:59]: Is authorization (user needs a role to read/create/update/delete) missing or are they handled other way or it's left for later?

**Review by pwargulak — CHANGES_REQUESTED** [review]
Left some questions

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java (hunk 1 above)
Comment (sradziszewski) [inline src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java:60]: As we put all the logic in the controller, not the service, maybe consider writing some unit tests to make sure the logic in the methods works is correct.

File: src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java (hunk 3 above)
Comment (sradziszewski) [inline src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java:59]: As we put all the logic in the controller, not the service, maybe consider writing some unit tests to make sure the logic in the methods works is correct.

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+ */
+@Controller
+@RequestMapping(DatasetController.RESOURCE_PATH)
+@Transactional
+public class DatasetController extends BaseController {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatasetController.class);
+
+  public static final String RESOURCE_PATH = ServerController.RESOURCE_PATH
+          + "/{serverId}" + "/datasets";
```
[the remark this answers is not in the document]
Comment (anawrotsoldevelo) [reply src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java]: I moved them to single line

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java (hunk 1 above)
Comment (anawrotsoldevelo) [reply src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java:59]: We have created separate ticket for this (SELV3-577)

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java (hunk 2 above)
Comment (anawrotsoldevelo) [reply src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java]: Nice idea, changed in both dataset and server controller

File: src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java (hunk 3 above)
Comment (anawrotsoldevelo) [reply src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java:56]: Left for later in SELV3-577

File: src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java (hunk 1 above)
Comment (anawrotsoldevelo) [reply src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java:59]: I have tested these controllers in integration tests instead

File: src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java (hunk 3 above)
Comment (anawrotsoldevelo) [reply src/main/java/org/openlmis/integration/dhis2/web/server/ServerController.java:56]: I have tested these controllers in integration tests instead

**Review by sradziszewski — APPROVED** [review]

**Comment by anawrotsoldevelo** [comment]
I also reverted server password hashing and created separate ticket for password encryption/decryption using some sort of cipher instead of bcrypt

**Review by pwargulak — APPROVED** [review]
Look good

**Pull request merged**